### PR TITLE
Rework ui_dispatch tests to avoid need for Qt

### PR DESCRIPTION
**Pull request by mdickinson**
Since #1788, we have only one test module that makes use of the Qt event loop. That test module contains tests for the behaviour of handlers that use `dispatch='ui'` mechanism to redispatch off-thread notifications to the ui thread.

This PR reworks that test module, with some significant collateral damage along the way.

In detail:

- reworks that test module (`test_ui_notifiers`) to avoid the need for the Qt event loop; instead, it tests against a `ui_handler` based on asyncio, which redispatches to the running asyncio event loop
- adds a `get_ui_handler` counterpart to `set_ui_handler`, and exposes both functions in `traits.api`
- adds type hints for `get_ui_handler` and `set_ui_handler`
- removes two public module globals from `trait_notifiers`: `ui_handler` has been made private, while `ui_thread` is removed altogether
- fixes a bug where ui dispatch didn't do the right thing (PR #1740 was incomplete; this bug should have been caught at review time on that PR)
- makes another couple of drive-by cleanups, removing a very old check for `threading.local()` being a dict (which it hasn't been in living memory), and tidying up some uses of thread identity.

### CODE REVIEW

File: traits/trait_notifiers.py [hunk 1]
```diff
@@ -615,10 +614,10 @@ class FastUITraitChangeNotifyWrapper(TraitChangeNotifyWrapper):
     """
 
     def dispatch(self, handler, *args):
-        if threading.current_thread().ident == ui_thread:
+        if threading.current_thread() == threading.main_thread():
```
Comment (mdickinson) [inline traits/trait_notifiers.py:617]: This is the fix that should have been part of #1740.

File: traits/trait_notifiers.py [hunk 2]
```diff
@@ -153,11 +159,7 @@ def _get_handlers(self):
             thread.
         """
         thread_local = self.thread_local
-        if isinstance(thread_local, dict):
```
Comment (mdickinson) [inline traits/trait_notifiers.py:156]: Drive-by cleanup: this condition is never true in modern Python.

**Comment by mdickinson** [comment]
@flongford Do you have bandwidth for review? This is part of a longer-term goal of insulating Traits from changes in Qt / PySide, and eventually decoupling Traits and TraitsUI entirely.

File: traits/tests/test_ui_notifiers.py [hunk 3]
```diff
@@ -61,88 +50,162 @@ def on_foo_change(self, obj, name, old, new):
...
+            [(threading.main_thread(), (self.obj, "foo", 0, 3))],
+        )
+        self.assertEqual(self.obj.foo, 3)
 
 
-class TestMethodUINotifiers(BaseTestUINotifiers, unittest.TestCase):
-    """ Tests for dynamic notifiers with `dispatch='ui'` set by method call.
-    """
+class TestMethodUINotifiers(
+    BaseTestUINotifiers, unittest.IsolatedAsyncioTestCase
```
Comment (flongford) [inline traits/tests/test_ui_notifiers.py:195]: Huh, TIL about `unittest.IsolatedAsyncioTestCase`!

File: traits/tests/test_ui_notifiers.py [hunk 4]
```diff
@@ -61,88 +50,162 @@ def on_foo_change(self, obj, name, old, new):
...
-    )
-    def test_notification_from_separate_thread(self):
+        # Then the notification is processed synchronously on the main thread.
+        self.assertEqual(
+            self.notifications,
+            [(threading.main_thread(), (self.obj, "foo", 0, 3))],
+        )
 
-        obj = self.obj_factory()
+    def test_notification_from_separate_thread_failure_case(self):
```
Comment (flongford) [inline traits/tests/test_ui_notifiers.py:157]: I think I understand the context for this test case, but just to be clear: do we accept that any background process unknown to the UI _will_ be able to modify objects, but ideally we also want to know about it?

If so, is this a new edge case that we can just test more easily without Qt, or does this happen often?

**Review by flongford — APPROVED** [review]
LGTM!

Thanks for the explanation of the code changes @mdickinson. The approach to separate the concept of "a ui handler" from "Qt's ui handler" looks like what we want here. The `asyncio` package also seems like an appropriate choice for creating a a test "ui handler".

I could mostly follow along with the what's being changed, though I've left some comments for my own understanding.

File: traits/trait_notifiers.py [hunk 5]
```diff
@@ -615,10 +614,10 @@ class FastUITraitChangeNotifyWrapper(TraitChangeNotifyWrapper):
     """
 
     def dispatch(self, handler, *args):
-        if threading.current_thread().ident == ui_thread:
+        if threading.current_thread() == threading.main_thread():
             handler(*args)
         else:
-            ui_handler(handler, *args)
+            _ui_handler(handler, *args)
```
[the remark this answers is not in the document]
Comment (mdickinson) [reply traits/trait_notifiers.py:620]: Yes, we definitely should! Now I have to go figure out why the tests aren't exercising this ...

File: traits/trait_notifiers.py [hunk 6]
```diff
@@ -27,27 +27,33 @@
...
 
 def set_ui_handler(handler):
     """ Sets up the user interface thread handler.
     """
-    global ui_handler, ui_thread
+    global _ui_handler
 
-    ui_handler = handler
-    ui_thread = threading.current_thread().ident
+    _ui_handler = handler
```
[the remark this answers is not in the document]
Comment (mdickinson) [reply traits/trait_notifiers.py:47]: Yes: TraitsUI already knows about and uses `set_ui_handler`. The `get_ui_handler` is really just for tests right now (mostly so that we can reset state at the end of a test that changes the handler), but it seems like an obvious counterpart to `set_ui_handler`.

File: traits/tests/test_ui_notifiers.py (hunk 4 above)
Comment (mdickinson) [reply traits/tests/test_ui_notifiers.py:157]: So the key point of the test (and of the `dispatch="ui"` functionality) is that even though the object is modified off the main thread, the trait change _handlers_ are run in the main UI thread.

That said, this pattern isn't ideal in the first place; my preference is to use something like Traits Futures so that we're not modifying the same object from multiple threads in the first place.

Not sure whether this answers your question or not ... ?

Comment (flongford) [reply traits/tests/test_ui_notifiers.py:157]: Ah, I see - so the main result of this test is that we don't see any notifiers and we observe an exception, since the handlers are always on the main UI tread.

That helps understand the context, yes, thanks!

File: traits/trait_notifiers.py (hunk 5 above)
[the remark this answers is not in the document]
Comment (mdickinson) [reply traits/trait_notifiers.py:620]: Hmm. I think they _are_ exercising this, but our tests for the exception don't distinguish between a `RuntimeError` with a nice message and a `TypeError` from trying to use `None` as a callable. I need to re-raise exceptions and catch them ...

[the remark this answers is not in the document]
Comment (mdickinson) [reply traits/trait_notifiers.py:622]: Reworked in [cc8653e](https://github.com/enthought/traits/pull/1792/commits/cc8653e5288f4087da8e26e30156afe24e8c2cf6)

**Comment by mdickinson** [comment]
@flongford I've responded to comments above and made one minor update, for better exception handling.

File: traits/tests/test_ui_notifiers.py (hunk 4 above)
Comment (mdickinson) [reply traits/tests/test_ui_notifiers.py:156]: Right - the notifiers don't get executed at all in this case.

**Comment by mdickinson** [comment]
> * removes two public module globals from `trait_notifiers`: `ui_handler` has been made private

This should not have been done: Pyface directly accesses `ui_handler`. (It shouldn't, but it does.)
